pkg/kubectl/cmd/exec.go

```diff
@@ -184,58 +185,64 @@ func (p *ExecOptions) Run() error {
 		containerName = pod.Spec.Containers[0].Name
 	}
 
-	// TODO: refactor with terminal helpers from the edit utility once that is merged
-	var stdin io.Reader
+	// ensure we can recover the terminal while attached
+	t := term.TTY{Parent: p.InterruptParent}
+
+	// check for TTY
 	tty := p.TTY
 	if p.Stdin {
-		stdin = p.In
-		if tty {
-			if file, ok := stdin.(*os.File); ok {
-				inFd := file.Fd()
-				if term.IsTerminal(inFd) {
-					oldState, err := term.SetRawTerminal(inFd)
-					if err != nil {
-						glog.Fatal(err)
-					}
-					// this handles a clean exit, where the command finished
-					defer term.RestoreTerminal(inFd, oldState)
-
-					// SIGINT is handled by term.SetRawTerminal (it runs a goroutine that listens
-					// for SIGINT and restores the terminal before exiting)
-
-					// this handles SIGTERM
-					sigChan := make(chan os.Signal, 1)
-					signal.Notify(sigChan, syscall.SIGTERM)
-					go func() {
-						<-sigChan
-						term.RestoreTerminal(inFd, oldState)
-						os.Exit(0)
-					}()
-				} else {
-					fmt.Fprintln(p.Err, "STDIN is not a terminal")
-				}
-			} else {
-				tty = false
-				fmt.Fprintln(p.Err, "Unable to use a TTY - input is not the right kind of file")
-			}
+		t.In = p.In
+		if tty && !t.IsTerminal() {
+			tty = false
+			fmt.Fprintln(p.Err, "Unable to use a TTY - input is not a terminal or the right kind of file")
+		}
+	}
+	t.Raw = tty
+
+	var resizeChan chan term.Size
+	if tty {
+		// make it buffered so we can send the initial terminal size without blocking, prior to starting
+		// the streaming below
+		resizeChan = make(chan term.Size, 1)
```

Comment:
Same comment about resizing - this feels like it could be moved to `pkg/util/term` as a helper that is cleanly separated (as the PR evolves)